src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -2455,33 +2424,105 @@ public override BoundNode VisitMethodGroup(BoundMethodGroup node)
             return null;
         }
 
-        public override BoundNode VisitNullCoalescingOperator(BoundNullCoalescingOperator node)
+#nullable enable
+
+        public override BoundNode? VisitNullCoalescingOperator(BoundNullCoalescingOperator node)
         {
-            VisitRvalue(node.LeftOperand);
             if (IsConstantNull(node.LeftOperand))
             {
-                VisitRvalue(node.RightOperand);
+                VisitRvalue(node.LeftOperand);
+                Visit(node.RightOperand);
             }
             else
             {
-                var savedState = this.State.Clone();
+                TLocalState savedState = VisitPossibleConditionalAccess(node.LeftOperand, out var stateWhenNotNull)
+                    ? stateWhenNotNull
+                    : State.Clone();
+
                 if (node.LeftOperand.ConstantValue != null)
                 {
                     SetUnreachable();
                 }
-                VisitRvalue(node.RightOperand);
-                Join(ref this.State, ref savedState);
+                Visit(node.RightOperand);
+                if (IsConditionalState)
+                {
+                    Join(ref StateWhenTrue, ref savedState);
+                    Join(ref StateWhenFalse, ref savedState);
+                }
+                else
+                {
+                    Join(ref this.State, ref savedState);
+                }
             }
             return null;
         }
 
-        public override BoundNode VisitConditionalAccess(BoundConditionalAccess node)
+        /// <summary>
+        /// Unconditionally visits an expression.
+        /// If the expression has "state when not null" after visiting,
+        /// the method returns 'true' and writes the state to <paramref name="stateWhenNotNull" />.
+        /// </summary>
+        protected bool VisitPossibleConditionalAccess(BoundExpression node, [NotNullWhen(true)] out TLocalState? stateWhenNotNull)
+        {
+            EnterRegionIfNeeded(node);
+            var hasStateWhenNotNull = visit(out stateWhenNotNull);
+            Debug.Assert(!IsConditionalState);
+            LeaveRegionIfNeeded(node);
+            return hasStateWhenNotNull;
+
+            bool visit([NotNullWhen(true)] out TLocalState? stateWhenNotNull)
+            {
+                var access = node switch
+                {
+                    BoundConditionalAccess ca => ca,
+                    BoundConversion { Conversion: Conversion innerConversion, Operand: BoundConditionalAccess ca } when isAcceptableConversion(ca, innerConversion) => ca,
```

Comment:
This doesn't look correct. I don't believe we should be walking through conversions here at all: the conversion that needs to be applied in in `BoundNullCoalescingOperator.LeftConversion`. The main question here is whether we want to handle cases like `((C)c1?.M2(x = 0)) ?? c1.Value.M3(x = 0)`. Do we want to handle that case?

Reply:
And if we _do_, I'm not sure we need to have a concept of an "acceptable" conversion. Either there was an error during initial binding, or the result of the conversion can be nullable.

---
In reply to: [597987186](https://github.com/dotnet/roslyn/pull/51567#discussion_r597987186) [](ancestors = 597987186)

Reply:
> Do we want to handle that case?

I think it's good to handle casts on conditional expressions, but if it proves more expensive/fraught, it might be necessary to drop it.

> And if we do, I'm not sure we need to have a concept of an "acceptable" conversion. Either there was an error during initial binding, or the result of the conversion can be nullable.

I didn't understand this part. It feels like what's important about the conversion is:

- If the conversion only gives non-null output when the input is non-null, we can propagate out the StateWhenNull (most built-in conversions meet this requirement).
- Otherwise, we don't know why the conversion result is non-null, and we have to discard the StateWhenNotNull of the inner conditional access.

> This doesn't look correct. I don't believe we should be walking through conversions here at all: the conversion that needs to be applied in in BoundNullCoalescingOperator.LeftConversion.

In https://github.com/dotnet/roslyn/pull/51567#discussion_r597876642 I wrote:
> After reading the spec, it feels like the implicit conversion on the left doesn't affect this analysis, since the null test always occurs before the left conversion. We just have to demonstrate that in the tests. #Closed

Did I get this wrong? Could you please give an example of when this analysis needs to account for the LeftConversion on the null coalescing operator?

Reply:
> Could you please give an example of when this analysis needs to account for the LeftConversion on the null coalescing operator?

It never does, that's the point I was making.

Looking through the tests, I think I see what your intent here is, I was misreading.